substrate/frame/support/procedural/src/pallet/parse/tests/mod.rs

```diff
@@ -0,0 +1,259 @@
+///
+/// The reason this must be specified is that our underlying parsing of pallets depends on
+/// reaching out into the file system to look for particular `Cargo.toml` dependencies via the
+/// [`generate_access_from_frame_or_crate`] method, so to simulate this properly in a proc
+/// macro crate, we need to temporarily convince this function that we are running from the
+/// directory of a valid pallet.
+#[macro_export]
+macro_rules! assert_pallet_parses {
+	(
+		#[manifest_dir($manifest_dir:literal)]
+		$($tokens:tt)*
+	) => {
+		{
+			let mut pallet: Option<$crate::pallet::parse::Def> = None;
+			$crate::pallet::parse::tests::simulate_manifest_dir($manifest_dir, core::panic::AssertUnwindSafe(|| {
+				pallet = Some($crate::pallet::parse::Def::try_from(syn::parse_quote! {
+					$($tokens)*
+				}, false).unwrap());
+			}));
+			pallet.unwrap()
+		}
+	}
+}
+
+/// Similar to [`assert_pallet_parses`], except this instead expects the pallet not to parse,
+/// and allows you to specify a regex matching the expected parse error.
+///
+/// This is identical syntactically to [`assert_pallet_parses`] in every way except there is a
+/// second attribute that must be specified immediately after `#[manifest_dir(..)]` which is
+/// `#[error_regex(..)]` which should contain a string/regex literal designed to match what you
+/// consider to be the correct parsing error we should see when we try to parse this particular
+/// pallet.
+///
+/// ## Example:
+///
+/// ```
+/// assert_pallet_parse_error! {
+/// 	#[manifest_dir("../../examples/basic")]
+/// 	#[error_regex("Missing `\\#\\[pallet::pallet\\]`")]
+/// 	#[frame_support::pallet]
+/// 	pub mod pallet {
+/// 		#[pallet::config]
+/// 		pub trait Config: frame_system::Config {}
+/// 	}
+/// }
+/// ```
+#[macro_export]
+macro_rules! assert_pallet_parse_error {
+	(
+		#[manifest_dir($manifest_dir:literal)]
+		#[error_regex($reg:literal)]
+		$($tokens:tt)*
+	) => {
+		$crate::pallet::parse::tests::simulate_manifest_dir($manifest_dir, || {
+			$crate::assert_error_matches!(
+				$crate::pallet::parse::Def::try_from(
+					parse_quote! {
+						$($tokens)*
+					},
+					false
+				),
+				$reg
+			);
+		});
+	}
+}
+
+/// Safely runs the specified `closure` while simulating an alternative `CARGO_MANIFEST_DIR`,
+/// restoring `CARGO_MANIFEST_DIR` to its original value upon completion regardless of whether
+/// the closure panics.
+///
+/// This is useful in tests of `Def::try_from` and other pallet-related methods that internally
+/// make use of [`generate_access_from_frame_or_crate`], which is sensitive to entries in the
+/// "current" `Cargo.toml` files.
+///
+/// This function uses a [`Mutex`] to avoid a race condition created when multiple tests try to
+/// modify and then restore the `CARGO_MANIFEST_DIR` ENV var in an overlapping way.
+pub fn simulate_manifest_dir<P: AsRef<std::path::Path>, F: FnOnce() + std::panic::UnwindSafe>(
+	path: P,
+	closure: F,
+) {
+	use std::{env::*, path::*};
```

Comment:
I think we should define the Mutex here, because this is the only place that we are using it and it is private anyway.